### PR TITLE
control pod restart - TEST-40381, 40382, 40386

### PR DESCRIPTION
**Pull request by rahul-hatwar**
Signed-off-by: Rahul Hatwar <rahulchandrakant.hatwar@seagate.com>

# Problem Statement
- Problem statement

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
-  [ ] New/Affected tests are executed on Latest Build
-  [ ] Attach test execution logs
-  [ ] Collection tested and no collection error introduced (`pytest --local True --collect-only`)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
-  [ ] If change in any common function, make sure to update all calls and execute all affected tests.

# Documentation
  Checklist for Author
-  [ ] Changes done to ReadMe / WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

**Comment by rahul-hatwar** [comment]
'TEST-32731', 'marks': ['s3_ops']}, {'nodeid': 'tests/s3/versioning/test_versioning_get_head_object.py::TestVersioningGetHeadObject::test_get_head_object_versioned_bucket_32729', 'test_id': 'TEST-32729', 'marks': ['s3_ops']}, {'nodeid': 'tests/s3/versioning/test_versioning_get_head_object.py::TestVersioningGetHeadObject::test_get_head_versioned_object_invalid_32727', 'test_id': 'TEST-32727', 'marks': ['s3_ops']}, {'nodeid': 'tests/s3/versioning/test_versioning_get_head_object.py::TestVersioningGetHeadObject::download_and_check', 'test_id': '', 'marks': []}, {'nodeid': 'tests/s3/versioning/test_versioning_put_object.py::TestVersioningPutObject::test_put_object_preexisting_32724', 'test_id': 'TEST-32724', 'marks': ['s3_ops']}, {'nodeid': 'tests/s3/versioning/test_versioning_put_object.py::TestVersioningPutObject::test_put_object_versioning_enabled_32728', 'test_id': 'TEST-32728', 'marks': ['s3_ops']}, {'nodeid': 'tests/s3/versioning/test_versioning_put_object.py::TestVersioningPutObject::test_put_object_versioning_suspended_32733', 'test_id': 'TEST-32733', 'marks': ['s3_ops']}, {'nodeid': 'tests/security/test_cortx_port_scanner_kubectl_svc.py::test_cortx_port_scanner_kubectl_svc', 'test_id': 'TEST-34217', 'marks': ['security']}, {'nodeid': 'tests/security/test_cortx_port_scanner_netstat.py::test_cortx_port_scanner_netstat', 'test_id': 'TEST-34218', 'marks': ['security']}] created at /root/rahul/cortx-test-1/log/te_meta.json
Successfully unmounted directory

=================================================== 2153 tests collected in 6.00s ===================================================
(virenv) [root@ssc-vm-g3-rhev4-2524 cortx-test-1]#

File: tests/ha/test_control_pod_restart.py [hunk 1]
```diff
@@ -729,3 +732,452 @@ def test_mpu_after_ctrl_pod_failover(self):
...
+            resp = output.get(timeout=HA_CFG["common_params"]["60sec_delay"])
+            if isinstance(resp, bool):
+                break
+
+        if resp is None:
+            assert_utils.assert_true(False, "Background process of chunk upload failed")
+
+        LOGGER.info("Step 5: Successfully verified response of background process")
+
+        if not resp:
```
Comment (ankita-93) [inline tests/ha/test_control_pod_restart.py:1134]: same as above. There is not failure expected in IOs

File: tests/ha/test_control_pod_restart.py [hunk 2]
```diff
@@ -729,3 +732,452 @@ def test_mpu_after_ctrl_pod_failover(self):
...
+        LOGGER.info("Step 7: Successfully downloaded object and verified checksum")
+
+        if CMN_CFG["dtm0_disabled"]:
+            LOGGER.info("Step 8: Create IAM user, buckets and upload objects on degraded ")
+            users = self.mgnt_ops.create_account_users(nusers=1)
+            self.test_prefix = 'test-40386-1'
+            self.s3_clean.update(users)
+            resp = self.ha_obj.ha_s3_workload_operation(s3userinfo=list(users.values())[0],
+                                                        log_prefix=self.test_prefix,
+                                                        nsamples=10)
```
Comment (ankita-93) [inline tests/ha/test_control_pod_restart.py:1174]: nsamples is already 10. no need to add here

File: tests/ha/test_control_pod_restart.py [hunk 3]
```diff
@@ -729,3 +732,452 @@ def test_mpu_after_ctrl_pod_failover(self):
...
+        else:
+            LOGGER.info("Perform IOs with variable object sizes ")
+            resp = self.ha_obj.ha_s3_workload_operation(s3userinfo=self.s3_clean,
+                                                        log_prefix=self.test_prefix,
+                                                        nsamples=10, skipcleanup=True)
```
Comment (ankita-93) [inline tests/ha/test_control_pod_restart.py:1179]: same

File: tests/ha/test_control_pod_restart.py [hunk 4]
```diff
@@ -729,3 +732,452 @@ def test_mpu_after_ctrl_pod_failover(self):
...
+        mpu_id = responses[3]
+        LOGGER.debug("Responses received from background process:\nexp_failed_parts: "
+                     "%s\nfailed_parts: %s\nparts_etag: %s\nmpu_id: %s", exp_failed_parts,
+                     failed_parts, parts_etag, mpu_id)
+        if len(exp_failed_parts) == 0 and len(failed_parts) == 0:
+            LOGGER.info("All the parts are uploaded successfully")
+        elif failed_parts:
+            assert_utils.assert_true(False, "Failed to upload parts when cluster was in degraded "
+                                            f"state. Failed parts: {failed_parts}")
+        elif exp_failed_parts:
```
[the remark this answers is not in the document]
Comment (ApurwaMohite) [reply tests/ha/test_control_pod_restart.py:831]: Correct.

File: tests/ha/test_control_pod_restart.py [hunk 5]
```diff
@@ -729,3 +735,416 @@ def test_mpu_after_ctrl_pod_failover(self):
...
+                'put_etag': put_etag}
+        thread = threading.Thread(target=self.ha_obj.create_bucket_copy_obj, args=(event,),
+                                  kwargs=args)
+        thread.daemon = True  # Daemonize thread
+        thread.start()
+        LOGGER.info("Step 2: Successfully started background process for copy object")
+        # While loop to sync this operation with background thread to achieve expected scenario
+        LOGGER.info("Waiting for creation of %s buckets", bkt_cnt)
+        bkt_list = list()
+        timeout = time.time() + HA_CFG["common_params"]["60sec_delay"]
```
Comment (ApurwaMohite) [inline tests/ha/test_control_pod_restart.py:936]: This should be 180 sec delay.
Please use HA_CFG["common_params"]["bucket_creation_delay"] for the delay.

File: tests/ha/test_control_pod_restart.py [hunk 6]
```diff
@@ -1289,3 +1293,410 @@ def test_copy_obj_after_ctrl_pod_failover(self):
...
+        LOGGER.info("Control pod %s is hosted on %s node", self.control_pod_name,
+                    self.control_node)
+
+        failover_node = self.system_random.choice([ele for ele in self.host_worker_list if ele !=
+                                                   self.control_node])
+        LOGGER.debug("Fail over node is: %s", failover_node)
+
+        LOGGER.info("Step 2: Failover control pod %s to node %s and check cluster status",
+                    self.control_pod_name, failover_node)
+        pod_yaml = {self.control_pod_name: self.modified_yaml}
```
Comment (ApurwaMohite) [inline tests/ha/test_control_pod_restart.py:1355]: Where event is set and cleared?

Comment (ApurwaMohite) [inline tests/ha/test_control_pod_restart.py:1355]: Where event is set and cleared?

File: tests/ha/test_control_pod_restart.py [hunk 7]
```diff
@@ -1289,3 +1293,410 @@ def test_copy_obj_after_ctrl_pod_failover(self):
...
+        LOGGER.info("Control pod %s is hosted on %s node", self.control_pod_name,
+                    self.control_node)
+
+        failover_node = self.system_random.choice([ele for ele in self.host_worker_list if ele !=
+                                                   self.control_node])
+        LOGGER.debug("Fail over node is: %s", failover_node)
+
+        LOGGER.info("Step 3: Failover control pod %s to node %s and check cluster status",
+                    self.control_pod_name, failover_node)
+        pod_yaml = {self.control_pod_name: self.modified_yaml}
```
Comment (ApurwaMohite) [inline tests/ha/test_control_pod_restart.py:1514]: Same as above.

Comment (ApurwaMohite) [inline tests/ha/test_control_pod_restart.py:1514]: Same as above.

File: tests/ha/test_control_pod_restart.py [hunk 8]
```diff
@@ -1289,3 +1293,410 @@ def test_copy_obj_after_ctrl_pod_failover(self):
...
```
Comment (ApurwaMohite) [inline tests/ha/test_control_pod_restart.py:1647]: Same as above.

File: tests/ha/test_control_pod_restart.py [hunk 9]
```diff
@@ -1289,3 +1293,414 @@ def test_copy_obj_after_ctrl_pod_failover(self):
...
+                    HA_CFG["common_params"]["30sec_delay"])
+        time.sleep(HA_CFG["common_params"]["30sec_delay"])
+
+        LOGGER.info("Control pod %s is hosted on %s node", self.control_pod_name,
+                    self.control_node)
+
+        failover_node = self.system_random.choice([ele for ele in self.host_worker_list if ele !=
+                                                   self.control_node])
+        LOGGER.debug("Fail over node is: %s", failover_node)
+        event.set()
```
Comment (ApurwaMohite) [inline tests/ha/test_control_pod_restart.py:1647]: In chunk upload case event is not needed as we are not passing any event. We can not differentiate chunks uploaded when event is set and when event was cleared.

File: tests/ha/test_control_pod_restart.py [hunk 10]
```diff
@@ -1289,3 +1293,412 @@ def test_copy_obj_after_ctrl_pod_failover(self):
...
+        LOGGER.info("Download object response: %s", resp)
+        assert_utils.assert_true(resp[0], resp[1])
+        download_checksum = self.ha_obj.cal_compare_checksum(file_list=[download_path],
+                                                             compare=False)[0]
+        assert_utils.assert_equal(upload_checksum, download_checksum,
+                                  f"Failed to match checksum: {upload_checksum},"
+                                  f" {download_checksum}")
+        LOGGER.info("Step 6: Successfully downloaded object and verified checksum")
+
+        LOGGER.info("Step 7: Create IAM user, buckets and upload objects on"
```
Comment (ankita-93) [inline tests/ha/test_control_pod_restart.py:1693]: after control pod restart

File: tests/ha/test_control_pod_restart.py [hunk 11]
```diff
@@ -1289,3 +1293,412 @@ def test_copy_obj_after_ctrl_pod_failover(self):
...
+                    " control pod restart ")
+        users = self.mgnt_ops.create_account_users(nusers=1)
+        self.test_prefix = 'test-40386-1'
+        self.s3_clean.update(users)
+        resp = self.ha_obj.ha_s3_workload_operation(s3userinfo=list(users.values())[0],
+                                                    log_prefix=self.test_prefix,
+                                                    nclients=2, nsamples=2)
```
Comment (ankita-93) [inline tests/ha/test_control_pod_restart.py:1700]: since this one is not degraded cluster, we can keep default worklaod only

**Review by ApurwaMohite — APPROVED** [review]

File: tests/ha/test_control_pod_restart.py [hunk 12]
```diff
@@ -1289,3 +1293,411 @@ def test_copy_obj_after_ctrl_pod_failover(self):
...
+        failed_parts = responses[1]
+        parts_etag = responses[2]
+        mpu_id = responses[3]
+        LOGGER.debug("Responses received from background process:\nexp_failed_parts: "
+                     "%s\nfailed_parts: %s\nparts_etag: %s\nmpu_id: %s", exp_failed_parts,
+                     failed_parts, parts_etag, mpu_id)
+        assert_utils.assert_true(len(failed_parts) != 0, "Failed to upload parts during "
+                                                         "control pod failover. Failed "
+                                                         f"parts: {failed_parts}")
+        assert_utils.assert_true(len(exp_failed_parts) != 0, "Failed to upload "
```
Comment (sonalk0209) [inline tests/ha/test_control_pod_restart.py:1393]: isn't this is when we are doing failover?

File: tests/ha/test_control_pod_restart.py [hunk 13]
```diff
@@ -1289,3 +1293,411 @@ def test_copy_obj_after_ctrl_pod_failover(self):
...
+        time.sleep(HA_CFG["common_params"]["20sec_delay"])
+
+        LOGGER.info("Control pod %s is hosted on %s node", self.control_pod_name,
+                    self.control_node)
+
+        failover_node = self.system_random.choice([ele for ele in self.host_worker_list if ele !=
+                                                   self.control_node])
+        LOGGER.debug("Fail over node is: %s", failover_node)
+        event.set()
+        LOGGER.info("Step 3: Failover control pod %s to node %s and check cluster status",
```
Comment (sonalk0209) [inline tests/ha/test_control_pod_restart.py:1513]: same as above

File: tests/ha/test_control_pod_restart.py [hunk 14]
```diff
@@ -1289,3 +1293,411 @@ def test_copy_obj_after_ctrl_pod_failover(self):
...
+        put_etag = responses[0]
+        exp_fail_bkt_obj_dict = responses[1]
+        failed_bkts = responses[2]
+        LOGGER.debug("Responses received from background process:\nput_etag: "
+                     "%s\nexp_fail_bkt_obj_dict: %s\nfailed_bkts: %s", put_etag,
+                     exp_fail_bkt_obj_dict, failed_bkts)
+        assert_utils.assert_true(len(failed_bkts) != 0, "Failed to do copy object when control "
+                                                        "pod restart. Failed "
+                                                        f"buckets: {failed_bkts}")
+        assert_utils.assert_true(len(exp_fail_bkt_obj_dict), f"Failed buckets "
```
Comment (sonalk0209) [inline tests/ha/test_control_pod_restart.py:1552]: same as above

**Review by sonalk0209 — COMMENTED** [review]
@rahul-hatwar plz check comments

File: tests/ha/test_control_pod_restart.py [hunk 15]
```diff
@@ -1289,3 +1293,411 @@ def test_copy_obj_after_ctrl_pod_failover(self):
...
+        time.sleep(HA_CFG["common_params"]["60sec_delay"])
+
+        LOGGER.info("Control pod %s is hosted on %s node", self.control_pod_name,
+                    self.control_node)
+
+        failover_node = self.system_random.choice([ele for ele in self.host_worker_list if ele !=
+                                                   self.control_node])
+        LOGGER.debug("Fail over node is: %s", failover_node)
+        event.set()
+        LOGGER.info("Step 2: Failover control pod %s to node %s and check cluster status",
```
[the remark this answers is not in the document]
Comment (rahul-hatwar) [reply tests/ha/test_control_pod_restart.py:1353]: We are checking the failover node frm 1346 to 1351, and the logging that in step 2, That's the reason it is outside

File: tests/ha/test_control_pod_restart.py (hunk 14 above)
Comment (rahul-hatwar) [reply tests/ha/test_control_pod_restart.py:1552]: In this also no expected bucket object failed, that's why assert.

**Review by ankita-93 — APPROVED** [review]

File: tests/ha/test_control_pod_restart.py [hunk 16]
```diff
@@ -1289,3 +1293,414 @@ def test_copy_obj_after_ctrl_pod_failover(self):
...
+            assert_utils.assert_true(False, "Background process failed to do multipart upload")
+
+        exp_failed_parts = responses[0]
+        failed_parts = responses[1]
+        parts_etag = responses[2]
+        mpu_id = responses[3]
+        LOGGER.debug("Responses received from background process:\nexp_failed_parts: "
+                     "%s\nfailed_parts: %s\nparts_etag: %s\nmpu_id: %s", exp_failed_parts,
+                     failed_parts, parts_etag, mpu_id)
+        assert_utils.assert_true(len(failed_parts) != 0, "Failed to upload parts during "
```
Comment (ankita-93) [inline tests/ha/test_control_pod_restart.py:1390]: This one is not during parts. This contains list of parts when event not set
We can change the satement accordingly.

**Review by sonalk0209 — APPROVED** [review]
LGTM

**Pull request merged**